azure-iot-device/azure/iot/device/provisioning/pipeline/pipeline_stages_provisioning.py

```diff
@@ -44,3 +58,388 @@ def _run_op(self, op):
+
+                        self_weakref = weakref.ref(self)
+
+                        @pipeline_thread.invoke_on_pipeline_thread_nowait
+                        def do_retry_after():
+                            this = self_weakref()
+                            logger.info(
+                                "{stage_name}({op_name}): retrying".format(
+                                    stage_name=this.name, op_name=op.name
+                                )
+                            )
+                            initial_register_op.retry_after_timer.cancel()
+                            initial_register_op.retry_after_timer = None
+                            initial_register_op.completed = False
+                            this.run_op(initial_register_op)
+
+                        logger.warning(
+                            "{stage_name}({op_name}): Op needs retry with interval {interval} because of {error}.  Setting timer.".format(
+                                stage_name=self.name,
+                                op_name=op.name,
+                                interval=retry_interval,
+                                error=error,
+                            )
+                        )
+
+                        initial_register_op.retry_after_timer = Timer(
+                            retry_interval, do_retry_after
+                        )
+                        initial_register_op.retry_after_timer.start()
+
+                    # Service success scenario
+                    else:
+                        registration_status = self._get_registration_status(decoded_response)
+                        if registration_status == "assigned" or registration_status == "failed":
+                            # process complete response here
+                            # TODO Is there an response when service returns "failed" ?
+                            complete_registration_result = self._form_complete_result(
+                                operation_id=operation_id,
+                                decoded_response=decoded_response,
+                                status=registration_status,
+                            )
+                            initial_register_op.registration_result = complete_registration_result
+                            if registration_status == "failed":
+                                error = exceptions.ServiceError(
+                                    "Registration operation returned failed registration status  with a status code of {status_code}".format(
+                                        status_code=success_status_code
+                                    )
+                                )
+                            # complete the original op in both cases
+                            initial_register_op.complete(error=error)
+                        elif registration_status == "assigning":
+
+                            def copy_result_to_original_op(op, error):
```

Comment:
What happens if the client has to query twice before it goes from assigning to assigned?  i don't see how this code handles the case where the SendQueryRequestOperation returns 429 or "assigning"

Reply:
are you referring to the code on line 166 or something else. That code is present in `on_query_response` function.

---
In reply to: [364413575](https://github.com/Azure/azure-iot-sdk-python/pull/432#discussion_r364413575) [](ancestors = 364413575)